File: android/app/src/main/java/com/rightpoint/lite_chat/IM/Msg.java
package com.rightpoint.lite_chat.IM;

import androidx.annotation.StringDef;

import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.util.HashMap;

/**
 * Description：
 * @author Wonder Wei
 * Create date：2020/7/24 9:31 AM 
 */
public class Msg extends HashMap<String, Object> {
    public static final String TYPE_TXT = "type_txt";
    public static final String TYPE_IMG = "type_img";
    public static final String TYPE_VOICE = "type_voice";
    public static final String TYPE_VIDEO = "type_video";
    public static final String TYPE_VIDEO_CALL = "type_video_call";

    @StringDef({TYPE_TXT, TYPE_IMG, TYPE_VOICE, TYPE_VIDEO, TYPE_VIDEO_CALL})
    @Retention(RetentionPolicy.SOURCE)
    public @interface MsgType {

    }

    public static final String CMD_ACTION = "videoCall";

    /**
     * 设置所属会话对方用户名
     * @param username 用户名，可为空
     * @return 消息体
     */
    public Msg setUsername(String username) {
        put("username", username);
        return this;
    }

    public String getUsername() {
        return (String) get("username");
    }

    public Msg setType(@MsgType String type) {
        put("type", type);
        return this;
    }

    public MsgType getType() {
        return (MsgType) get("type");
    }

    public Msg setFrom(String from) {
        put("from", from);
        return this;
    }

    public String getFrom() {
        return (String) get("from");
    }

    public Msg setTo(String to) {
        put("to", to);
        return this;
    }

    public String getTo() {
        return (String) get("to");
    }

    public Msg setTime(long time) {
        put("time", time);
        return this;
    }

    public long getTime() {
        return (long) get("time");
    }

    public Msg setTxt(String txt) {
        put("txt", txt);
        return this;
    }

    public String getTxt() {
        return (String) get("txt");
    }

    public Msg setOriginal(boolean original) {
        put("original", original);
        return this;
    }

    public boolean getOriginal() {
        return (boolean) get("original");
    }

    public Msg setThumbUrl(String thumbUri) {
        put("thumbUrl", thumbUri);
        return this;
    }

    public String getThumbUrl() {
        return (String) get("thumbUrl");
    }

    public Msg setWidth(int width) {
        put("width", width);
        return this;
    }

    public int getWidth() {
        return (int) get("width");
    }

    public Msg setHeight(int height) {
        put("height", height);
        return this;
    }

    public int getHeight() {
        return (int) get("height");
    }

    public Msg setImgUrl(String imgUri) {
        put("imgUrl", imgUri);
        return this;
    }

    public String getImgUrl() {
        return (String) get("imgUrl");
    }

    public Msg setVoiceUri(String voiceUri) {
        put("voiceUri", voiceUri);
        return this;
    }

    public String getVoiceUri() {
        return (String) get("voiceUri");
    }

    public Msg setLength(int length) {
        put("length", length);
        return this;
    }

    public int getLength() {
        return (int) get("length");
    }

    public Msg setVideoUri(String videoUri) {
        put("videoUri", videoUri);
        return this;
    }

    public String getVideoUri() {
        return (String) get("videoUri");
    }

    public Msg setFileUri(String fileUri) {
        put("fileUri", fileUri);
        return this;
    }

    public String getFileUri() {
        return (String) get("fileUri");
    }
}

File: android/app/src/main/java/com/rightpoint/lite_chat/IM/Msg.kt
package com.rightpoint.lite_chat.IM

import androidx.annotation.StringDef
import java.util.*

/**
 * Description：
 * @author Wonder Wei
 * Create date：2020/7/24 9:31 AM
 */
class Msg : HashMap<String?, Any?>() {
    companion object {
        const val TYPE_TXT = "type_txt"
        const val TYPE_IMG = "type_img"
        const val TYPE_VOICE = "type_voice"
        const val TYPE_VIDEO = "type_video"
        const val TYPE_VIDEO_CALL = "type_video_call"
        const val CMD_ACTION = "videoCall"
    }
    
    @StringDef(TYPE_TXT, TYPE_IMG, TYPE_VOICE, TYPE_VIDEO, TYPE_VIDEO_CALL)
    @Retention(AnnotationRetention.SOURCE)
    annotation class MsgType

    /**
     * 设置所属会话对方用户名
     * @param username 用户名，可为空
     * @return 消息体
     */
    fun setUsername(username: String?): Msg {
        put("username", username)
        return this
    }

    fun getUsername() = get("username") as String?

    fun setType(@MsgType type: String): Msg {
        put("type", type)
        return this
    }

    fun getType() = get("type") as MsgType?

    fun setFrom(from: String?): Msg {
        put("from", from)
        return this
    }

    fun getFrom() = get("from") as String?

    fun setTo(to: String?): Msg {
        put("to", to)
        return this
    }

    fun getTo() = get("to") as String?

    fun setTime(time: Long): Msg {
        put("time", time)
        return this
    }

    fun getTime() = get("time") as Long

    fun setTxt(txt: String?): Msg {
        put("txt", txt)
        return this
    }

    fun getTxt() = get("txt") as String?

    fun setOriginal(original: Boolean): Msg {
        put("original", original)
        return this
    }

    fun getOriginal() = get("original") as Boolean

    fun setThumbUrl(thumbUri: String?): Msg {
        put("thumbUrl", thumbUri)
        return this
    }

    fun getThumbUrl() = get("thumbUrl") as String?

    fun setWidth(width: Int): Msg {
        put("width", width)
        return this
    }

    fun getWidth() = get("width") as Int

    fun setHeight(height: Int): Msg {
        put("height", height)
        return this
    }

    fun getHeight() = get("height") as Int

    fun setImgUrl(imgUri: String?): Msg {
        put("imgUrl", imgUri)
        return this
    }

    fun getImgUrl() = get("imgUrl") as String?

    fun setVoiceUri(voiceUri: String?): Msg {
        put("voiceUri", voiceUri)
        return this
    }

    fun getVoiceUri() = get("voiceUri") as String?

    fun setLength(length: Int): Msg {
        put("length", length)
        return this
    }

    fun getLength() = get("length") as Int

    fun setVideoUri(videoUri: String?): Msg {
        put("videoUri", videoUri)
        return this
    }

    fun getVideoUri() = get("videoUri") as String?

    fun setFileUri(fileUri: String?): Msg {
        put("fileUri", fileUri)
        return this
    }

    fun getFileUri() = get("fileUri") as String?
}